File: arrays/in_place_operations/move_zeroes/move_zeroes.cpp
/**
 * Author: Nathan Bockisch
 * Date: October 2, 2021
 **/

class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int beg_ptr = 0;
        
        // Remove zeros and maintian relative order
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i]) {
                nums[beg_ptr++] = nums[i];
            }
        }
        
        // Write zeros back in to the end of array
        for (; beg_ptr < nums.size(); nums[beg_ptr++] = 0);
    }
};

File: array_and_string/conclusion/move_zeroes/move_zeroes.rs
/*
 * Author: Nathan Bockisch
 * Date: July 15, 2022
 */
impl Solution {
    pub fn move_zeroes(nums: &mut Vec<i32>) {
        let (mut slow, mut fast) = (0, 1);

        while fast < nums.len() {
            if nums[slow] == 0 && nums[fast] != 0 {
                nums.swap(slow, fast);
            }

            if nums[slow] != 0 { slow += 1 }
            fast += 1;
        }
    }
}
